### ISSUE TITLE
Latest set of changes seriously screwed up the thread pool

### ISSUE THREAD

**Issue by Aaronontheweb** [helios-io/DedicatedThreadPool#10]
cc @rogeralsing 

Not sure what we did exactly, but it's become trivially easy to cause massive latencies in the benchmark with the dedicated threadpool since the last PR. I'm talking 3 seconds to go through 10000 items. That bad.

When we switched to the following:

``` csharp
Workers = Enumerable.Repeat(0, settings.NumThreads).Select(_ => new WorkerQueue()).ToArray();
```

Nearly every benchmark run since has either deadlocked or slowed to a crawl

**Comment by Aaronontheweb** [comment]
BTW @rogeralsing, the code you added is definitely correct - but it exposed a bug in our earlier code.

**Comment by Aaronontheweb** [comment]
So I've been able to fix the deadlock issue by getting rid of all of the manual reset events and replacing the `ConcurrentQueue` with a `BlockingCollection`.

However, the `ThreadPool` now outperforms us. This means that the previous buggy implementation, where all workers shared the same queue was actually much more performant, oddly enough.

Right way to solve this problem is to figure out the dead lock issue. However, I've spent too much time on this today - I need to move onto the dispatcher stuff. Going to ship the blocking queue version in a NuGet release and wait for an update.
